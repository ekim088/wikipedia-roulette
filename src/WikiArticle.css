.articles .wa {
	display: inline-block;
	margin: 0 40px 0 0;
}

.articles .wa:last-child {
	margin-right: 0;
}

.wa {
	border-radius: 8px;
	box-shadow: 0 10px 30px -8px #bec1bd;
	color: #070707;
	font-family: "proxima-nova", sans-serif;
	font-weight: 600;
	line-height: 150%;
	margin: 0 auto;
	overflow: hidden;
	position: relative;
	transition: all .3s ease-in-out;
}

.wa-body {
	align-items: center;
	background-image: linear-gradient(135deg, rgba(115,103,240,.95) 50%, rgba(206,159,252,.85) 100%);
	box-sizing: border-box;
	color: #fff;
	-ms-flex-direction: column;
	flex-direction: column;
	display: -ms-flexbox;
	display: flex;
	height: 300px;
	padding: 30px;
	position: relative;
	max-width: 500px;
}

.wa-content {
	margin: auto;
}

.wa h2 {
	font-family: "basic-sans", sans-serif;
	font-size: 34px;
	font-weight: 400;
	line-height: 110%;
	margin: 0 0 20px;
}

.wa h3 {
	font-family: "basic-sans", sans-serif;
	font-size: 14px;
	font-weight: 700;
	letter-spacing: .5px;
	line-height: 140%;
	margin: 0 0 14px;
	text-transform: uppercase;
}

.wa h2 + h3 {
	margin-bottom: 20px;
	margin-top: -10px;
}

.wa-img {
	background-color: #e6e7e8;
	background-position: center;
	background-repeat: no-repeat;
	background-size: cover;
	filter: blur(1.5px);
	height: 100%;
	left: 0;
	overflow: hidden;
	position: absolute;
	top: 0;
	width: 100%;
}
