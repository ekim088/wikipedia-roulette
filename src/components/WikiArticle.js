// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import trim from '../utils/trim';
import getArticleSummary from '../utils/wikipediaHandler';
import type { ArticleResponse } from '../utils/wikipediaHandler';
import '../scss/_WikiArticle.scss';
import defaultImg from '../default-article-image.jpg';

import addArticle from '../actions/addArticle';

// flow types
type Article = {
	description: ?string,
	externalUrl?: string,
	id: ?string,
	image?: string,
	summary: ?string,
	title: ?string
};

export type SharedArticle = {
	title: ?string,
	description: ?string
};

export type Props = {
	...Article,
	callback?: SharedArticle => any,
	dispatchArticle?: SharedArticle => void
};

type State = {
	...Article,
	isLoaded: boolean
};

/**
 * Parses raw article summary response to format consumable by Component.
 * @param {Object} summaryResponse Article summary response.
 * @returns {Object} Parsed article data.
 */
export const parseArticleSummary = ({
	content_urls: contentUrls,
	description,
	extract,
	pageid,
	thumbnail: { source } = {},
	title
}: ArticleResponse): Article => ({
	description,
	externalUrl:
		(contentUrls && contentUrls.desktop && contentUrls.desktop.page) || '',
	id: pageid,
	image: source || defaultImg,
	summary: extract,
	title
});

class WikiArticle extends Component<Props, State> {
	static defaultProps = {
		callback: undefined,
		externalUrl: undefined,
		dispatchArticle: undefined,
		image: undefined
	};

	constructor(props: Props) {
		super(props);

		const { description, externalUrl, id, image, summary, title } = props;
		this.state = {
			description,
			externalUrl,
			id,
			image,
			summary,
			title,
			isLoaded: !!(id && title && description && summary)
		};
	}

	async componentDidMount() {
		const { callback, dispatchArticle } = this.props;
		const { isLoaded } = this.state;

		// fetch Wikipedia article and update state if data not available
		if (!isLoaded) {
			await this.fetchArticle();
		}

		// apply article data to shared state
		const { title, description } = this.state;
		const sharedArticleData = { title, description };

		if (typeof dispatchArticle === 'function') {
			dispatchArticle(sharedArticleData);
		}

		// perform post-mount callback
		if (typeof callback === 'function') {
			callback(sharedArticleData);
		}
	}

	async fetchArticle() {
		const response = await getArticleSummary();
		const article: Article = { ...parseArticleSummary(response) };

		// update state with article data
		this.setState({
			...article,
			isLoaded: true
		});
	}

	render() {
		const {
			description,
			externalUrl,
			id,
			image,
			isLoaded,
			summary,
			title
		} = this.state;
		const summaryId: ?string = id ? `summary-${id}` : null;
		const titleId: ?string = id ? `title-${id}` : null;

		return (
			<article
				className={`wa${isLoaded ? '' : ' loading'} style-0`}
				data-id={id}
			>
				<div className={`wa-header${image ? ' has-image' : ''}`}>
					{image && (
						<div
							className="wa-img"
							style={{ backgroundImage: `url('${image}')` }}
							role="img"
							aria-labelledby={titleId}
							aria-describedby={summaryId}
						/>
					)}
					<div className="wa-header-content">
						{title && <h2 className="wa-title">{trim(title, 50)}</h2>}
						{description && (
							<h3 className="wa-description">{trim(description, 100)}</h3>
						)}
					</div>
				</div>
				<div className="wa-content">
					{summary && (
						<p className="wa-summary" id={summaryId}>
							{trim(summary, 250)}
						</p>
					)}
					<a href={externalUrl} target="_blank" rel="noopener noreferrer">
						Test Link
					</a>
				</div>
			</article>
		);
	}
}

const mapStateToProps = () => ({});
const mapDispatchToProps = dispatch => ({
	dispatchArticle: (data: SharedArticle): void => dispatch(addArticle(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(WikiArticle);
