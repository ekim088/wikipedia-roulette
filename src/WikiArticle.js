import React, { Component } from 'react';
import './WikiArticle.css';
import { WikiApiHandler } from './makeApiHandler';
import defaultImg from './default-article-image.jpg';

class WikiArticle extends Component {
	constructor(props) {
		super(props);
		this.state = {
			title: '',
			summary: '',
			image: defaultImg,
			callbackPosition: null
		};
	}

	componentWillMount() {
		/*
		 * define a globally accessible callback method Wiki API Handler is able
		 * to call to be able to update this component's state when JSONP request
		 * completes
		 */
		const callback = (response) => {
			const updateObj = {};

			// set up object with state values to update
			for (let key in response) {
				if (response.hasOwnProperty(key) &&
					response[key] &&
					this.state.hasOwnProperty(key)) {
					updateObj[key] = response[key];
				}
			}

			// update Component state
			if (Object.keys(updateObj).length > 0) {
				this.setState(updateObj);
			}

			// parse article content is available
			if (response.htmlStr) {
				this.parseArticleHtml(response.htmlStr);
			}
		};
		let callbackPosition;

		// push callback to global array
		callbackPosition = WikiApiHandler.articleCallbacks.length;
		this.setState({ callbackPosition: WikiApiHandler.articleCallbacks.length });
		WikiApiHandler.articleCallbacks.push(callback);

		// initiate call to API
		WikiApiHandler.getArticleFromComponent(callbackPosition);
	}

	/**
	 * Parses HTML string of Wikipedia content to fit customized format.
	 * 
	 * @param {string} htmlStr 
	 */
	parseArticleHtml(htmlStr) {
		/**
		 * Parses DOM element for summary content.
		 * 
		 * @param {Element} el Element containing article content
		 */
		const parseHtmlForSummary = (el) => {
			// max length of summary content
			const lengthMax = 300;

			if (typeof el !== 'object') {
				console.log(`[WikiArticle] Invalid object - 
					could not parse article content for summary`);
				return;
			}

			// parse HTML content for article summary
			// currently just grabs first paragraph contents
			console.log(`[WikiArticle] Parsing article content for summary`);
			let firstPar = el.querySelector('p:not(.mw-empty-elt)');

			if (firstPar) {
				let summaryStr = firstPar.innerText.length > lengthMax ?
					firstPar.innerText.slice(0, lengthMax) + '...' :
					firstPar.innerText;

				// update article with summary
				this.setState({ summary: summaryStr });
			}
		};

		/**
		 * Parses DOM element for image content.
		 * 
		 * @param {Element} el Element containing article content
		 */
		const parseHtmlForImage = (el) => {
			const dimensionThreshold = 39999;
			let images = html.querySelectorAll('img');

			if (typeof el !== 'object') {
				console.log(`[WikiArticle] Invalid object - 
					could not parse article content for image`);
				return;
			}

			console.log(`[WikiArticle] Parsing article content for image`);

			// search image for one large enough to use as main article graphic
			// compares images dimensions against a threshold
			for (let image of images) {
				let height = image.getAttribute('height');
				let width = image.getAttribute('width');

				if (height && width && (height * width) > dimensionThreshold) {
					// update article with image
					let src = image.getAttribute('src');
					console.log(`[WikiArticle] Using ${src} as image for ${this.state.title}`);
					this.setState({ image: src });
					return;
				}
			}

			// if image not found in article, call Wikipedia API to attempt to grab a related image
			if (typeof this.state.title !== 'undefined' && 
				typeof this.state.callbackPosition !== 'undefined') {
				WikiApiHandler.getImageFromComponent(this.state.title, this.state.callbackPosition);
			}
		};

		// create element with HTML string contents
		let html = document.createElement('div');
		html.innerHTML = htmlStr;

		// parse contents
		parseHtmlForSummary(html);
		parseHtmlForImage(html);

		// remove ref to temp element; does this actually matter?
		html = undefined;
	}

	render() {
		return (
			<div className="WikiArticle">
				<div className="content">
					<h2 className="title">{this.state.title}</h2>
					<p className="summary">{this.state.summary}</p>
				</div>
				<div
					className="graphic"
					style={{
						backgroundImage: `url('${this.state.image}')`
					}}
				>
				</div>
			</div>
		);
	}
}

export default WikiArticle;
