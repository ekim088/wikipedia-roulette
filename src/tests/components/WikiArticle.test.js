import WikiArticle, { parseArticleSummary } from '../../components/WikiArticle';
import { mockWikipediaData, wikipediaHandlerTestUtil } from '../mocks';

describe('WikiArticle', () => {
	let container;
	let wikipediaHandlerMock;

	beforeEach(() => {
		container = document.createElement('div');
		document.body.appendChild(container);
		wikipediaHandlerMock = wikipediaHandlerTestUtil.mock();
	});

	afterEach(() => {
		document.body.removeChild(container);
		container = null;

		// reset mocks
		wikipediaHandlerTestUtil.reset();
	});

	it('renders without crashing', async () => {
		ReactDOM.render(
			<Provider store={configureStore()}>
				<WikiArticle />
			</Provider>,
			container
		);

		// wait for component to finish async calls
		await act(async () => {});

		ReactDOM.unmountComponentAtNode(container);
	});

	it('renders an article and fetches article data from Wikipedia if not provided', async () => {
		const wrapper = mount(
			<Provider store={configureStore()}>
				<WikiArticle />
			</Provider>
		);

		// wait for component to finish async calls and update state
		await act(async () => {});
		wrapper.update();

		expect(wikipediaHandlerMock).toHaveBeenCalled();
		expect(wrapper).toMatchSnapshot();
	});

	it('should not fetch article data if provided', async () => {
		const wrapper = mount(
			<Provider store={configureStore()}>
				<WikiArticle
					id="someId"
					title="Some Title"
					description="Some description"
					summary="Some summary."
				/>
			</Provider>
		);

		// wait for component to finish async calls and update state
		await act(async () => {});
		wrapper.update();

		expect(wikipediaHandlerMock).not.toHaveBeenCalled();
		expect(wrapper).toMatchSnapshot();
	});

	it('should translate the article response data on parseArticleSummary', () => {
		expect(parseArticleSummary(mockWikipediaData)).toEqual({
			description: 'some description',
			externalUrl: 'someDesktopUrl.com',
			id: 'someId',
			image: 'someimage.png',
			summary: 'some extract',
			title: 'Some Title'
		});
	});
});
