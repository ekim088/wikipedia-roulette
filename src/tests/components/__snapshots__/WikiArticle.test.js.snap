// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WikiArticle renders an article and fetches article data from Wikipedia if not provided 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(WikiArticle)>
    <WikiArticle
      dispatchArticle={[Function]}
    >
      <article
        className="wa style-0"
        data-id="someId"
      >
        <div
          className="wa-header has-image"
        >
          <div
            aria-describedby="summary-someId"
            aria-labelledby="title-someId"
            className="wa-img"
            role="img"
            style={
              Object {
                "backgroundImage": "url('someimage.png')",
              }
            }
          />
          <div
            className="wa-header-content"
          >
            <h2
              className="wa-title"
            >
              Some Title
            </h2>
            <h3
              className="wa-description"
            >
              some description
            </h3>
          </div>
        </div>
        <div
          className="wa-content"
        >
          <p
            className="wa-summary"
            id="summary-someId"
          >
            some extract
          </p>
          <a
            href="someDesktopUrl.com"
            rel="noopener noreferrer"
            target="_blank"
          >
            Test Link
          </a>
        </div>
      </article>
    </WikiArticle>
  </Connect(WikiArticle)>
</Provider>
`;

exports[`WikiArticle should not fetch article data if provided 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(WikiArticle)
    description="Some description"
    id="someId"
    summary="Some summary."
    title="Some Title"
  >
    <WikiArticle
      description="Some description"
      dispatchArticle={[Function]}
      id="someId"
      summary="Some summary."
      title="Some Title"
    >
      <article
        className="wa style-0"
        data-id="someId"
      >
        <div
          className="wa-header"
        >
          <div
            className="wa-header-content"
          >
            <h2
              className="wa-title"
            >
              Some Title
            </h2>
            <h3
              className="wa-description"
            >
              Some description
            </h3>
          </div>
        </div>
        <div
          className="wa-content"
        >
          <p
            className="wa-summary"
            id="summary-someId"
          >
            Some summary.
          </p>
          <a
            rel="noopener noreferrer"
            target="_blank"
          >
            Test Link
          </a>
        </div>
      </article>
    </WikiArticle>
  </Connect(WikiArticle)>
</Provider>
`;
